### ISSUE TITLE
[css-values-5] add definition of <dimension-unit>

### ISSUE THREAD

**Issue by SofyaPovarova** [w3c/csswg-drafts#9648]
This fix requires #9646 and #9647 to be done at first. Can we add the definition of [<dimension-unit>](https://drafts.csswg.org/css-values-5/#typedef-dimension-unit)?
`<dimension-unit> = '%' | <length-unit> | <angle-unit> | <time-unit> | <frequency-unit> | <flex-unit>`

**Comment by Loirooriol** [comment]
You can see the diagram in https://drafts.csswg.org/css-syntax-3/#dimension-token-diagram
The unit of a `<dimension-token>` is just an `<ident-token>`.
Your proposal is missing things like https://drafts.csswg.org/css-variables-2/#variable-units

**Comment by Loirooriol** [comment]
Ok I was a bit confused given the linkes issues where you want redefine `<length>` and others.
This is not about redefining `<dimension>`, it's just the `<dimension-unit>` used for `attr()`, currently defined in prose, right?

**Comment by tabatkins** [comment]
Can you elaborate on why you need a more specific definition for `<dimension-unit>`? The current definition is perfectly well-defined, it's just in prose rather than written directly in grammar.

**Comment by SofyaPovarova** [comment]
Indeed, you're right. Thank you for pointing that out. It seems that '%' should not be included in the definition of \<dimension-unit\>, as it is currently [done here](https://drafts.csswg.org/css-values-5/#typedef-dimension-unit) and should be:
`<dimension-unit> = <length-unit> | <angle-unit> | <time-unit> | <frequency-unit> | <flex-unit>`

\<dimension-unit\> is used only for \<attr-type\> which can be '%'. Therefore,  \<attr-type\> should be fixed too,  '%' should be added to \<attr-type\>: 
`<attr-type> = string | url | ident | color | number | percentage | length | angle | time | frequency | flex | % | <dimension-unit>`

I agree, variable unit is definitely important but it feels like handling that might be a whole separate task.
Just to note, I'm relatively new to reading and fixing specifications, there's a chance I might miss something simple.

Could it be that this section in the specification is quite recent, and it is better to hold off on creating an issue for it at this point?

**Comment by SofyaPovarova** [comment]
We are exploring the possibility of using [Webref](https://github.com/w3c/webref) JSONs for automated validation in JetBrains IDEs.
Webref has JSONs with scraped from specification definitions in the main branch and also JSONs with manual fixed data in the curated branch. Some entities don't have value definitions, there were not only primitive values, but also a list of values like \<named-color\> or \<dimension-unit\> which I mentioned here. I'm working on finding a way to handle these values and the most effective approach seems to be adding definitions if feasible.
Initially, I addressed a couple of issues with examples to the Webref repo and was directed towards improving the specifications themselves, which seems to be the right approach. While I understand that it might pose some challenges, I think it's essential to at least raise these cases and investigate potential solutions

**Comment by tabatkins** [comment]
> We are exploring the possibility of using [Webref](https://github.com/w3c/webref) JSONs for automated validation in JetBrains IDEs.

Okay. In general, we do not write the specs with the *intention* of them being machine-consumeable; they're intended to be human-readable, instead. That means that some grammar productions end up with prose definitions even if they *could* be written purely in Value Definition Syntax, because it's clearer for humans and less of an editing hazard.

For example, if we were to explicitly define `<dimension-unit>` in the grammar, as `px | em| ...`, then we would have to remember to adjust it every time we added a new unit to CSS. It's very easy to forget to do so, and have the definitions drift from correctness. With its current prose definition, tho, it's always correct.

> Although now I wonder why the prose definition of `<dimension-unit>` contains '%' if `<dimension-token>` from css-syntax-3 can't consume '%'. I think it'll be better to define dimension-unit without '%', should it be fixed? 

No, it's good as-is. The fact that percentages are technically a different kind of token in the parser is an implementation detail; for *authors*, percentages and dimensions are identical. And in the particular use-case here, where the behavior is basically just "parse the attribute as a number, then smack the unit onto it", percentages and units are indeed exactly the same for authors.

(We do basically the exact same thing in Typed OM, where [`CSSUnitValue`](https://drafts.css-houdini.org/css-typed-om/#simple-numeric) is used to express all dimensions, percentages, and plain numbers. For the latter two, the `.unit` property is just `"percent"` or `"number"`. Again, this is because the distinction between these three separate token types is irrelevant for authors, so we don't reproduce the distinction in JS.)

**Comment by cdoublev** [comment]
For reference, `<dimension-unit>` has been renamed to `<attr-unit>` between 95e6ec6 and c4a4a61, and the open range is justified in 7fa0d2f.

Not sure if this issue can be closed though.
